src/test/coins_tests.cpp

```diff
@@ -147,8 +141,17 @@ BOOST_AUTO_TEST_CASE(coins_cache_simulation_test)
         {
             uint256 txid = txids[InsecureRandRange(txids.size())]; // txid we're going to modify in this iteration.
             Coin& coin = result[COutPoint(txid, 0)];
+
+            bool test_havecoin = InsecureRandBits(3) == 0;
+            bool result_havecoin = test_havecoin ? stack.back()->HaveCoin(COutPoint(txid, 0)) : false;
             const Coin& entry = (InsecureRandRange(500) == 0) ? AccessByTxid(*stack.back(), txid) : stack.back()->AccessCoin(COutPoint(txid, 0));
             BOOST_CHECK(coin == entry);
+            if (test_havecoin) BOOST_CHECK(result_havecoin == !entry.IsSpent());
+
+            if (InsecureRandBits(4) == 0) {
```

Comment:
What's the point of the two InsecureRandBits calls skipping the new BOOST_CHECKs? It seems like the checks should always pass and aren't very expensive. Is the point to avoid creating cache entries during the test? Maybe add a comment saying why this only randomly checks HaveCoin results.